tests/gen/gen_predefined.go

```diff
@@ -0,0 +1,202 @@
+// Copyright 2023 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package gen
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/client/request"
+)
+
+// GeneratePredefinedFromSDL generates documents for a schema from a predefined list
+// of docs that might include nested docs.
+// The schema is parsed to get the list of fields, and the docs
+// are created with the fields parsed from the schema.
+// This allows us to have only one large list of docs with predefined
+// fields, and create schemas with different fields from it.
+func GeneratePredefinedFromSDL(gqlSDL string, docsList DocsList) ([]GeneratedDoc, error) {
+	resultDocs := make([]GeneratedDoc, 0, len(docsList.Docs))
+	typeDefs, err := parseSchema(gqlSDL)
+	if err != nil {
+		return nil, err
+	}
+	generator := docGenerator{types: typeDefs}
+	for _, doc := range docsList.Docs {
+		docs, err := generator.GenerateDocs(doc, docsList.ColName)
+		if err != nil {
+			return nil, err
+		}
+		resultDocs = append(resultDocs, docs...)
+	}
+	return resultDocs, nil
+}
+
+// GeneratePredefined generates documents from a predefined list
+// of docs that might include nested docs.
+func GeneratePredefined(defs []client.CollectionDefinition, docsList DocsList) ([]GeneratedDoc, error) {
+	resultDocs := make([]GeneratedDoc, 0, len(docsList.Docs))
+	typeDefs := make(map[string]client.CollectionDefinition)
+	for _, col := range defs {
+		typeDefs[col.Description.Name] = col
+	}
+	generator := docGenerator{types: typeDefs}
+	for _, doc := range docsList.Docs {
+		docs, err := generator.GenerateDocs(doc, docsList.ColName)
+		if err != nil {
+			return nil, err
+		}
+		resultDocs = append(resultDocs, docs...)
+	}
+	return resultDocs, nil
+}
+
+type docGenerator struct {
+	types map[string]client.CollectionDefinition
+}
+
+func createDocJSON(typeDef *client.CollectionDefinition, doc map[string]any) string {
```

Comment:
todo: It is not obvious as to why this function is not just:
```
func createDocJSON(doc map[string]any) string{
   return json.Marshal(doc)
}
```
Can you please document what this function does and why we need to have custom json serialization logic here.

Reply:
there is no json handling in `gen` package anymore

Reply:
Nice - thanks Islam